File: 2196.create-binary-tree-from-descriptions.java
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

/*
 * @lc app=leetcode id=2196 lang=java
 *
 * [2196] Create Binary Tree From Descriptions
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {
    }

    TreeNode(int val) {
        this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution {
    public TreeNode createBinaryTree(int[][] descriptions) {
        Map<Integer, TreeNode> map = new HashMap<>();
        Set<Integer> set = new HashSet<>();
        for (int[] arr : descriptions) {
            int parent = arr[0], child = arr[1], isChild = arr[2];
            set.add(child);
            TreeNode node = map.getOrDefault(parent, new TreeNode(parent));
            if (isChild == 1) {
                node.left = map.getOrDefault(child, new TreeNode(child));
                map.put(child, node.left);
            } else {
                node.right = map.getOrDefault(child, new TreeNode(child));
                map.put(child, node.right);
            }
            map.put(parent, node);

        }

        int root = -1;
        for (int[] arr : descriptions) {
            if (!set.contains(arr[0])) {
                root = arr[0];
                break;
            }
        }

        return map.getOrDefault(root, null);

    }
}
// @lc code=end

File: python/2196.create-binary-tree-from-descriptions.py
#
# @lc app=leetcode id=2196 lang=python3
#
# [2196] Create Binary Tree From Descriptions
#

# @lc code=start
# Definition for a binary tree node.
from typing import List, Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def createBinaryTree(self, descriptions: List[List[int]]) -> Optional[TreeNode]:
        dict = {}
        setVar = set()

        for parent, child, childPointer in descriptions:

            setVar.add(child)
            parentNode = None
            parentNode = dict.get(
                parent) if parent in dict else TreeNode(parent)

            if childPointer == 1:
                childNode = dict.get(
                    child) if child in dict else TreeNode(child)
                parentNode.left = childNode
                dict[child] = childNode
            else:
                childNode = dict.get(
                    child) if child in dict else TreeNode(child)
                parentNode.right = childNode
                dict[child] = childNode

            dict[parent] = parentNode

        root = None
        for list in descriptions:
            if not list[0] in setVar:
                root = list[0]
                break

        return None if root is None else dict.get(root)

# @lc code=end
